### ISSUE TITLE
Phase 60.3: AttentionOptimizer — FlashAttention & Efficient Attention Patterns

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1132]
# Phase 60.3 — AttentionOptimizer

## Overview

The AttentionOptimizer implements hardware-aware attention computation strategies. Centered on Dao et al.'s FlashAttention algorithm, it provides IO-aware exact attention, multi-query/grouped-query attention variants, and sparse attention patterns for efficient long-context inference.

### FlashAttention Integration
- Tiling-based attention with O(N) memory instead of O(N²)
- SRAM-aware computation scheduling
- FlashAttention-2 with better parallelism and work partitioning
- FlashDecoding for inference (split-K attention for long sequences)
- Automatic fallback for unsupported head dimensions

### Multi-Query & Grouped-Query Attention
- MQA: single KV head shared across all query heads
- GQA: configurable KV head groups (e.g., 8 query heads → 2 KV groups)
- Automatic MHA → GQA conversion for pretrained models
- KV cache memory savings proportional to group ratio

### Linear Attention Alternatives
- Performer random feature attention (O(N) complexity)
- Linear Transformers with ELU feature maps
- Hybrid attention: linear for long-range, softmax for local
- Quality-latency tradeoff configuration

### Sparse Attention Patterns
- Longformer-style sliding window + global attention
- BigBird random + window + global attention
- Learned sparsity patterns from attention statistics
- Block-sparse attention with configurable block sizes
- Dynamic sparsity based on input characteristics

## Technical Specification

```typescript
interface AttentionConfig {
  algorithm: 'flash' | 'flash2' | 'standard' | 'linear';
  headType: 'mha' | 'mqa' | 'gqa';
  gqaGroups?: number;
  sparsePattern?: 'dense' | 'sliding' | 'longformer' | 'bigbird';
  windowSize?: number;
  globalTokens?: number[];
  blockSize: number;
}

interface AttentionOptimizer {
  computeAttention(query: Tensor, key: Tensor, value: Tensor, config: AttentionConfig): Tensor;
  convertToGQA(model: Model, numGroups: number): Model;
  profileAttention(inputLengths: number[]): AttentionProfile;
  selectOptimalConfig(constraints: LatencyConstraints): AttentionConfig;
  benchmarkConfigs(configs: AttentionConfig[]): BenchmarkResults;
}
```

## Dependencies
- Phase 60.1 (KVCacheManager) — KV cache layout compatibility
- Phase 57 (MoE) — attention within expert layers

## References
- Dao et al. — FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness
- Dao — FlashAttention-2: Faster Attention with Better Parallelism and Work Partitioning
- Shazeer — Fast Transformer Decoding: One Write-Head is All You Need (MQA)
- Ainslie et al. — GQA: Training Generalized Multi-Query Transformer Models
- Choromanski et al. — Rethinking Attention with Performers
- Beltagy et al. — Longformer: The Long-Document Transformer
- Zaheer et al. — Big Bird: Transformers for Longer Sequences

## Acceptance Criteria
- [ ] FlashAttention integration with automatic kernel selection
- [ ] GQA conversion preserving model quality (< 1% degradation)
- [ ] Sparse attention patterns with configurable sparsity levels
- [ ] 2-4x memory reduction for long sequences vs standard attention
- [ ] Automatic attention config selection based on hardware profile
- [ ] Unit tests with > 95% coverage

Depends on: #1130
Planning: #1129
